runtime/src/main/java/org/corfudb/runtime/view/stream/BackpointerStreamView.java

```diff
@@ -0,0 +1,367 @@
+            }
+
+            // Now, we do the actual write. We could get an overwrite
+            // exception here - any other exception we should pass up
+            // to the client.
+            try {
+                runtime.getAddressSpaceView()
+                        .write(tokenResponse.getToken(),
+                                Collections.singleton(streamID),
+                                object,
+                                tokenResponse.getBackpointerMap(),
+                                tokenResponse.getStreamAddresses());
+                // The write completed successfully, so we return this
+                // address to the client.
+                return tokenResponse.getToken();
+            } catch (OverwriteException oe) {
+                log.trace("Overwrite occurred at {}", tokenResponse);
+                // We got overwritten, so we call the deacquisition callback
+                // to inform the client we didn't get the address.
+                if (deacquisitionCallback != null) {
+                    if (!deacquisitionCallback.apply(tokenResponse)) {
+                        log.debug("Deacquisition requested abort");
+                        return -1L;
+                    }
+                }
+                // Request a new token, informing the sequencer we were
+                // overwritten.
+                tokenResponse = runtime.getSequencerView()
+                        .nextToken(Collections.singleton(streamID),
+                             1, true, false);
+            }
+        }
+    }
+
+    /** {@inheritDoc}
+     *
+     * The backpointer version of remaining() calls nextUpTo() multiple times,
+     * as it uses the default implementation in IStreamView. While this may
+     * appear to be non-optimized, these reads will most likely hit in the
+     * address space cache since the entries were read in order to resolve the
+     * backpointers.
+     * */
+    public synchronized LogData nextUpTo(long maxGlobal) {
+        // Pop the context if it has changed.
+        if (getCurrentContext().logPointer >=
+                getCurrentContext().maxAddress) {
+            StreamContext last = streamContexts.pollFirst();
+            log.trace("Completed context {}@{}, removing.",
+                    last.contextID, last.maxAddress);
+        }
+
+        // If we have no entries to read, fill the read queue.
+        if (getCurrentContext().readQueue.isEmpty()) {
+            fillReadQueue(maxGlobal);
+            // If we still have no entries to read, there's nothing in
+            // the stream (we linearized the read during fillReadQueue()).
+            if (getCurrentContext().readQueue.isEmpty()) {
+                return null;
+            }
+        }
+
+        // If the lowest value in the read queue is greater than
+        // max global, there's nothing for us to do.
+        if (getCurrentContext().readQueue.first() > maxGlobal) {
+            return null;
```

Comment:
why should it throw trimmed? So in the case where you have two threads:
```
T1: remainingUpTo(100); -> entry at address 0
T2: remainingUpTo(0);
```
You're saying T2 should throw trimmedException? Why? Nothing was trimmed?